intel-sgx/ias/src/verifier.rs

```diff
@@ -0,0 +1,751 @@
+        AutoDetect {}
+    }
+}
+
+impl Platform for AutoDetect {
+    fn verify(&self, for_self: bool, _nonce: &Option<String>, isv_enclave_quote_status: QuoteStatus, advisories: &Vec<IasAdvisoryId>) -> Result<(), Error> {
+        // The report must be for a valid quote.
+        match (for_self, isv_enclave_quote_status) {
+            (false, QuoteStatus::SwHardeningNeeded) => {
+                /* We don't know how the enclave peer has been compiled. Enclaves may verify their own
+                 * attestations, so we can ignore `QuoteStatus::SwHardeningNeeded` here and rely on the
+                 * logic of enclave versions. It is of paramount importance that we only ignore
+                 * `QuoteStatus::SwHardeningNeeded` cases.
+                 * `QuoteStatus::ConfigurationAndSwHardeningNeeded` means that the platform itself
+                 * needs to be updated as well. (Advisory IDs on how to do this will be included in the
+                 * `report.advisory_ids`) Such attestations should *always* be rejected. */
+                Ok(())
+            },
+            (true, QuoteStatus::SwHardeningNeeded) => {
+                /* Lets the enclave verify its own attestation. This enables us to inspect the compiler
+                 * used, and verify all software mitigations are in place. Quotes with status
+                 * `QuoteStatus::ConfigurationAndSwHardeningNeeded` represent insecure platforms and
+                 * should always be rejected. */
+                if advisories
+                    .iter()
+                    .any(|adv| !MITIGATED_SECURITY_ADVISORIES.contains(adv)) {
+                    let missing_ids = advisories
+                        .iter()
+                        .filter_map(|id|
+                            if MITIGATED_SECURITY_ADVISORIES.contains(id) {
+                                None
+                            } else {
+                                Some(id.as_str().to_string())
+                            })
+                        .collect::<Vec<_>>()
+                        .join(", ");
+                    Err(Error{
+                        error_kind: ErrorKind::ReportBadQuoteStatus(isv_enclave_quote_status, None),
+                        cause: Some(format!("Not all security advisories are mitigated. Missing mitigations: {}", missing_ids).into()),
+                    })
+                } else {
+                    Ok(())
+                }
+            },
+            (_, QuoteStatus::Ok) => {
+                Ok(())
+            },
+            _ => {
+                Err(Error{
+                    error_kind: ErrorKind::ReportBadQuoteStatus(isv_enclave_quote_status, None),
+                    cause: Some(format!("Quote status of {:?} is not trustworthy", isv_enclave_quote_status).into()),
+                })
+            },
+        }
+    }
+}
+
+impl VerifyAttestationEvidenceResponse<Unverified> {
+    #[cfg(feature = "manipulate_attestation")]
```

Comment:
Where is this used? In testing only?

Reply:
Yes, this can help in testing. It is checked in `lib.rs`:
```
#[cfg(all(feature = "manipulate_attestation", not(debug_assertions)))]                                                                                                                                                                      
compile_error!("The 'manipulate_attestation' feature is only available in debug mode."); 
```